encryptor/queryDataEncryptor.go

```diff
@@ -451,6 +447,43 @@ func (encryptor *QueryDataEncryptor) encryptUpdateValues(update *sqlparser.Updat
 	return encryptor.encryptValuesWithPlaceholders(values, placeholders, schema)
 }
 
+// updatePlaceholderMap matches the placeholder of a value to its column and records this into the mapping.
+func (encryptor *QueryDataEncryptor) updatePlaceholderMap(values []base.BoundValue, placeholders map[int]string, placeholder *sqlparser.SQLVal, columnName string) error {
+	// TODO(ilammy, 2020-10-15): handle MySQL placeholders too
+	// MySQL placeholders do not contain indices, you just need to count them
+	// and sequentially assign to values and columns.
+	switch placeholder.Type {
+	case sqlparser.PgPlaceholder:
+		// PostgreSQL placeholders look like "$1". Parse the number out of them.
+		text := string(placeholder.Val)
+		index, err := strconv.Atoi(strings.TrimPrefix(text, "$"))
+		if err != nil {
+			logrus.WithField("placeholder", text).WithError(err).Warning("Cannot parse placeholder")
+			return err
+		}
+		// Placeholders use 1-based indexing and "values" (Go slice) are 0-based.
+		index--
+		if index >= len(values) {
```

Comment:
hm, it is a questionable check. We should be sure that postgresql doesn't allow something like `insert into table1 values ($1+$2, $3, $4)`, where $1 and $2 may be integers (which should be summed) or strings (which should be concatenated). But such cases we can't support, but it is another error. We should warn that it is not an invalid query but unsupported feature.

Reply:
It does not check the columns, it checks the values. Effectively, that
```
prepare something as insert into table1 values ($1+$2, $3, $4)
```
is followed by
```
execute something(1, 2, 3, 4)
```
where there are enough values for all placeholders. If there are two values, for example, we just bail out with an error and skip this **Bind** packet. The database is likely to reply with an error too.

I'm not sure this is a right decision – to let the database see column values which should have been encrypted – but I don't feel that we should police the queries too hard and deny their processing if Acra cannot parse them.

In regards to that, I'd like to reserve a proper action for a later PR. It should improve error handling in such cases, and make query analysis more explicit. Then it will be possible to configure Acra behavior for queries which are too hard to analyze: either we block the query and report an error, or let the query through unchanged, or change whatever we can change and leave out the rest, or something else.